// @comunica/expression-evaluator/lib/evaluators/evaluatorHelpers/AsyncRecursiveEvaluator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncRecursiveEvaluator = void 0;
const E = require("../../expressions");
const TermTransformer_1 = require("../../transformers/TermTransformer");
const Err = require("../../util/Errors");
const BaseExpressionEvaluator_1 = require("./BaseExpressionEvaluator");
class AsyncRecursiveEvaluator extends BaseExpressionEvaluator_1.BaseExpressionEvaluator {
    constructor(context, termTransformer) {
        super(termTransformer ?? new TermTransformer_1.TermTransformer(context.superTypeProvider));
        this.context = context;
        this.subEvaluators = {
            // Shared
            [E.ExpressionType.Term]: (expr, _mapping) => this.term(expr),
            [E.ExpressionType.Variable]: (expr, mapping) => this.variable(expr, mapping),
            // Sync
            [E.ExpressionType.Operator]: (expr, mapping) => this.evalOperator(expr, mapping),
            [E.ExpressionType.SpecialOperator]: (expr, mapping) => this
                .evalSpecialOperator(expr, mapping),
            [E.ExpressionType.Named]: (expr, mapping) => this.evalNamed(expr, mapping),
            [E.ExpressionType.Existence]: (expr, mapping) => this.evalExistence(expr, mapping),
            [E.ExpressionType.Aggregate]: (expr, _mapping) => this.evalAggregate(expr),
            [E.ExpressionType.AsyncExtension]: (expr, mapping) => this.evalAsyncExtension(expr, mapping),
        };
    }
    async evaluate(expr, mapping) {
        const evaluator = this.subEvaluators[expr.expressionType];
        if (!evaluator) {
            throw new Err.InvalidExpressionType(expr);
        }
        return evaluator.bind(this)(expr, mapping);
    }
    async evalOperator(expr, mapping) {
        const argPromises = expr.args.map(arg => this.evaluate(arg, mapping));
        const argResults = await Promise.all(argPromises);
        return expr.apply(argResults);
    }
    async evalSpecialOperator(expr, mapping) {
        const evaluate = this.evaluate.bind(this);
        const context = {
            args: expr.args,
            mapping,
            superTypeProvider: this.context.superTypeProvider,
            now: this.context.now,
            baseIRI: this.context.baseIRI,
            functionArgumentsCache: this.context.functionArgumentsCache,
            evaluate,
            bnode: this.context.bnode,
            defaultTimeZone: this.context.defaultTimeZone,
        };
        return expr.applyAsync(context);
    }
    async _evalAsyncArgs(args, mapping) {
        const argPromises = args.map(arg => this.evaluate(arg, mapping));
        return await Promise.all(argPromises);
    }
    async evalNamed(expr, mapping) {
        return expr.apply(await this._evalAsyncArgs(expr.args, mapping));
    }
    async evalAsyncExtension(expr, mapping) {
        return await expr.apply(await this._evalAsyncArgs(expr.args, mapping));
    }
    async evalExistence(expr, mapping) {
        if (!this.context.exists) {
            throw new Err.NoExistenceHook();
        }
        return new E.BooleanLiteral(await this.context.exists(expr.expression, mapping));
    }
    // TODO: Remove?
    async evalAggregate(expr) {
        if (!this.context.aggregate) {
            throw new Err.NoAggregator();
        }
        return this.termTransformer.transformRDFTermUnsafe(await this.context.aggregate(expr.expression));
    }
}
exports.AsyncRecursiveEvaluator = AsyncRecursiveEvaluator;
