// @comunica/actor-query-source-identify-rdfjs/lib/ActorQuerySourceIdentifyRdfJs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorQuerySourceIdentifyRdfJs = void 0;
const bindings_factory_1 = require("@comunica/bindings-factory");
const bus_query_source_identify_1 = require("@comunica/bus-query-source-identify");
const core_1 = require("@comunica/core");
const QuerySourceRdfJs_1 = require("./QuerySourceRdfJs");
/**
 * A comunica RDFJS Query Source Identify Actor.
 */
class ActorQuerySourceIdentifyRdfJs extends bus_query_source_identify_1.ActorQuerySourceIdentify {
    constructor(args) {
        super(args);
    }
    async test(action) {
        const source = action.querySourceUnidentified;
        if (source.type !== undefined && source.type !== 'rdfjs') {
            throw new Error(`${this.name} requires a single query source with rdfjs type to be present in the context.`);
        }
        if (typeof source.value === 'string' || !('match' in source.value)) {
            throw new Error(`${this.name} received an invalid rdfjs query source.`);
        }
        return true;
    }
    async run(action) {
        return {
            querySource: {
                source: new QuerySourceRdfJs_1.QuerySourceRdfJs(action.querySourceUnidentified.value, await bindings_factory_1.BindingsFactory.create(this.mediatorMergeBindingsContext, action.context)),
                context: action.querySourceUnidentified.context ?? new core_1.ActionContext(),
            },
        };
    }
}
exports.ActorQuerySourceIdentifyRdfJs = ActorQuerySourceIdentifyRdfJs;
