// @comunica/expression-evaluator/lib/evaluators/SyncEvaluator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncEvaluator = void 0;
const lru_cache_1 = require("lru-cache");
const AlgebraTransformer_1 = require("../transformers/AlgebraTransformer");
const DateTimeHelpers_1 = require("../util/DateTimeHelpers");
const SyncRecursiveEvaluator_1 = require("./evaluatorHelpers/SyncRecursiveEvaluator");
class SyncEvaluator {
    static completeContext(context) {
        const now = context.now ?? new Date(Date.now());
        return {
            now,
            baseIRI: context.baseIRI ?? undefined,
            functionArgumentsCache: context.functionArgumentsCache ?? {},
            superTypeProvider: {
                cache: context.typeCache ?? new lru_cache_1.LRUCache({ max: 1_000 }),
                discoverer: context.getSuperType ?? (() => 'term'),
            },
            extensionFunctionCreator: context.extensionFunctionCreator,
            exists: context.exists,
            aggregate: context.aggregate,
            bnode: context.bnode,
            defaultTimeZone: context.defaultTimeZone ?? (0, DateTimeHelpers_1.extractTimeZone)(now),
        };
    }
    constructor(algExpr, context = {}) {
        this.algExpr = algExpr;
        this.context = context;
        // eslint-disable-next-line unicorn/no-useless-undefined
        const creator = context.extensionFunctionCreator ?? (() => undefined);
        const baseContext = SyncEvaluator.completeContext(context);
        const transformer = new AlgebraTransformer_1.AlgebraTransformer({
            type: 'sync',
            creator,
            ...baseContext,
        });
        this.expr = transformer.transformAlgebra(algExpr);
        this.evaluator = new SyncRecursiveEvaluator_1.SyncRecursiveEvaluator(baseContext, transformer);
    }
    evaluate(mapping) {
        const result = this.evaluator.evaluate(this.expr, mapping);
        return result.toRDF();
    }
    evaluateAsEBV(mapping) {
        const result = this.evaluator.evaluate(this.expr, mapping);
        return result.coerceEBV();
    }
    evaluateAsInternal(mapping) {
        return this.evaluator.evaluate(this.expr, mapping);
    }
}
exports.SyncEvaluator = SyncEvaluator;
