// @comunica/actor-query-result-serialize-stats/lib/ActorQueryResultSerializeStats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorQueryResultSerializeStats = void 0;
const bus_query_result_serialize_1 = require("@comunica/bus-query-result-serialize");
const asynciterator_1 = require("asynciterator");
const readable_stream_1 = require("readable-stream");
const process = require('process/');
/**
 * Serializes SPARQL results for testing and debugging.
 */
class ActorQueryResultSerializeStats extends bus_query_result_serialize_1.ActorQueryResultSerializeFixedMediaTypes {
    /* eslint-disable max-len */
    /**
     * @param args -
     *   \ @defaultNested {{ "stats": 0.5 }} mediaTypePriorities
     *   \ @defaultNested {{ "stats": "https://comunica.linkeddatafragments.org/#results_stats" }} mediaTypeFormats
     *   \ @defaultNested {<default_observer> a <caqrsst:components/ActionObserverHttp.jsonld#ActionObserverHttp>} httpObserver
     */
    constructor(args) {
        super(args);
    }
    /* eslint-enable max-len */
    async testHandleChecked(action, _context) {
        if (!['bindings', 'quads'].includes(action.type)) {
            throw new Error('This actor can only handle bindings streams or quad streams.');
        }
        return true;
    }
    pushHeader(data) {
        const header = ['Result', 'Delay (ms)', 'HTTP requests',
        ].join(',');
        data.push(`${header}\n`);
    }
    createStat(startTime, result) {
        const row = [result, this.delay(startTime), this.httpObserver.requests,
        ].join(',');
        return `${row}\n`;
    }
    createFooter(startTime) {
        const footer = ['TOTAL', this.delay(startTime), this.httpObserver.requests,
        ].join(',');
        return `${footer}\n`;
    }
    async runHandle(action, _mediaType, _context) {
        const data = new readable_stream_1.Readable();
        const resultStream = action.type === 'bindings' ?
            action.bindingsStream :
            action.quadStream;
        const startTime = this.now();
        let result = 1;
        function* end(cb) {
            yield cb();
        }
        const stream = (0, asynciterator_1.wrap)(resultStream)
            .map(() => this.createStat(startTime, result++))
            .append((0, asynciterator_1.wrap)(end(() => this.createFooter(startTime))));
        this.pushHeader(data);
        data.wrap(stream);
        return { data };
    }
    /* istanbul ignore next */
    now() {
        // TODO: remove when we will drop support of Node 14
        if (typeof performance === 'undefined') {
            const time = process.hrtime();
            return time[0] * 1_000 + (time[1] / 1_000_000);
        }
        return performance.now();
    }
    delay(startTime) {
        return this.now() - startTime;
    }
}
exports.ActorQueryResultSerializeStats = ActorQueryResultSerializeStats;
