// @comunica/bindings-factory/lib/BindingsFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BindingsFactory = void 0;
const immutable_1 = require("immutable");
const rdf_data_factory_1 = require("rdf-data-factory");
const Bindings_1 = require("./Bindings");
/**
 * A Bindings factory that provides Bindings backed by immutable.js.
 */
class BindingsFactory {
    constructor(dataFactory = new rdf_data_factory_1.DataFactory(), contextMergeHandlers) {
        this.dataFactory = dataFactory;
        this.contextMergeHandlers = contextMergeHandlers;
    }
    static async create(mediatorMergeBindingsContext, context) {
        return new BindingsFactory(new rdf_data_factory_1.DataFactory(), (await mediatorMergeBindingsContext.mediate({ context })).mergeHandlers);
    }
    bindings(entries = []) {
        return new Bindings_1.Bindings(this.dataFactory, (0, immutable_1.Map)(entries.map(([key, value]) => [key.value, value])), this.contextMergeHandlers ? { contextMergeHandlers: this.contextMergeHandlers } : undefined);
    }
    fromBindings(bindings) {
        return this.bindings([...bindings]);
    }
    fromRecord(record) {
        return this.bindings(Object.entries(record).map(([key, value]) => [this.dataFactory.variable(key), value]));
    }
}
exports.BindingsFactory = BindingsFactory;
