// @comunica/actor-context-preprocess-query-source-skolemize/lib/QuerySourceSkolemized.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuerySourceSkolemized = void 0;
const metadata_1 = require("@comunica/metadata");
const asynciterator_1 = require("asynciterator");
const utils_1 = require("./utils");
/**
 * A IQuerySource wrapper that skolemizes outgoing quads and bindings.
 */
class QuerySourceSkolemized {
    constructor(innerSource, sourceId) {
        this.innerSource = innerSource;
        this.sourceId = sourceId;
    }
    async getSelectorShape(context) {
        return this.innerSource.getSelectorShape(context);
    }
    queryBindings(operation, context, options) {
        const operationMapped = (0, utils_1.deskolemizeOperation)(operation, this.sourceId);
        if (!operationMapped) {
            const it = new asynciterator_1.ArrayIterator([], { autoStart: false });
            it.setProperty('metadata', {
                state: new metadata_1.MetadataValidationState(),
                cardinality: { type: 'exact', value: 0 },
                canContainUndefs: false,
                variables: [],
            });
            return it;
        }
        return (0, utils_1.skolemizeBindingsStream)(this.innerSource.queryBindings(operationMapped, context, options), this.sourceId);
    }
    queryBoolean(operation, context) {
        return this.innerSource.queryBoolean(operation, context);
    }
    queryQuads(operation, context) {
        const operationMapped = (0, utils_1.deskolemizeOperation)(operation, this.sourceId);
        if (!operationMapped) {
            const it = new asynciterator_1.ArrayIterator([], { autoStart: false });
            it.setProperty('metadata', {
                state: new metadata_1.MetadataValidationState(),
                cardinality: { type: 'exact', value: 0 },
            });
            return it;
        }
        return (0, utils_1.skolemizeQuadStream)(this.innerSource.queryQuads(operationMapped, context), this.sourceId);
    }
    queryVoid(operation, context) {
        return this.innerSource.queryVoid(operation, context);
    }
    get referenceValue() {
        return this.innerSource.referenceValue;
    }
    toString() {
        return `${this.innerSource.toString()}(SkolemID:${this.sourceId})`;
    }
}
exports.QuerySourceSkolemized = QuerySourceSkolemized;
