// @comunica/expression-evaluator/lib/aggregators/GroupConcat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupConcat = void 0;
const Helpers_1 = require("../functions/Helpers");
const Aggregator_1 = require("./Aggregator");
class GroupConcat extends Aggregator_1.AggregatorComponent {
    constructor() {
        super(...arguments);
        this.state = undefined;
        this.lastLanguageValid = true;
        this.lastLanguage = undefined;
    }
    static emptyValue() {
        return (0, Helpers_1.string)('').toRDF();
    }
    put(term) {
        if (this.state === undefined) {
            this.state = term.value;
            if (term.termType === 'Literal') {
                this.lastLanguage = term.language;
            }
        }
        else {
            this.state += this.separator + term.value;
            if (this.lastLanguageValid && term.termType === 'Literal' && this.lastLanguage !== term.language) {
                this.lastLanguageValid = false;
                this.lastLanguage = undefined;
            }
        }
    }
    result() {
        if (this.state === undefined) {
            return GroupConcat.emptyValue();
        }
        if (this.lastLanguageValid && this.lastLanguage) {
            return (0, Helpers_1.langString)(this.state, this.lastLanguage).toRDF();
        }
        return (0, Helpers_1.string)(this.state).toRDF();
    }
}
exports.GroupConcat = GroupConcat;
