// @comunica/expression-evaluator/lib/aggregators/Sum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sum = void 0;
const functions_1 = require("../functions");
const Helpers_1 = require("../functions/Helpers");
const C = require("../util/Consts");
const Aggregator_1 = require("./Aggregator");
class Sum extends Aggregator_1.AggregatorComponent {
    constructor() {
        super(...arguments);
        this.state = undefined;
        this.summer = functions_1.regularFunctions[C.RegularOperator.ADDITION];
    }
    static emptyValue() {
        return (0, Helpers_1.integer)(0).toRDF();
    }
    put(term) {
        if (this.state === undefined) {
            this.state = this.termToNumericOrError(term);
        }
        else {
            const internalTerm = this.termToNumericOrError(term);
            this.state = this.summer.apply([this.state, internalTerm], this.sharedContext);
        }
    }
    result() {
        if (this.state === undefined) {
            return Sum.emptyValue();
        }
        return this.state.toRDF();
    }
}
exports.Sum = Sum;
